torch/distributed/fsdp/flat_param.py

```diff
@@ -575,6 +585,402 @@ def shard_metadata(
             self.flat_param._shard_param_offsets[:],  # type: ignore[attr-defined]
         )
 
+    ###################
+    # UNSHARD/RESHARD #
+    ###################
+    def pre_unshard(self):
+        """
+        Postcondition: ``self.flat_param`` 's data is on the device for
+            communication and is what should be all-gathered. This means that
+            it matches the dtype of the expected unsharded parameter.
+        """
+        if self.uses_sharded_strategy and not self.needs_unshard():
+            # Do not prepare the sharded flattened parameter for an all-gather
+            # if the handle does not need to be unsharded
+            return
+        if self._uses_param_mixed_precision and not self._force_full_precision:
+            self._use_low_precision_shard()
+        elif self._config.offload_params and self.flat_param.device != self.device:
+            # NOTE: This creates a new tensor distinct from any attributes.
+            self._flat_param_to(self.device, non_blocking=True)
```

Comment:
looks like `to_cpu()` is not using pinned memory, I assume `non_blocking=True` is not making a difference here?